android/WALT/app/src/main/java/org/chromium/latency/walt/HistogramChart.java

```diff
@@ -0,0 +1,158 @@
+
+import java.text.DecimalFormat;
+import java.util.ArrayList;
+
+public class HistogramChart extends BarChart {
+
+    private static final float BIN_WIDTH = 5;
+    private final ArrayList<ArrayList<Double>> rawData;
+    private final ArrayList<IBarDataSet> dataSets;
+
+    private int minBin = 0;
+    private int maxBin = 100;
+    private double min = 0;
+    private double max = 100;
+
+    public HistogramChart(Context context, AttributeSet attrs) {
+        super(context, attrs);
+
+        final TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.HistogramChart);
+        final String descString;
+        final int numDataSets;
+        try {
+            descString = a.getString(R.styleable.HistogramChart_description);
+            numDataSets = a.getInteger(R.styleable.HistogramChart_numDataSets, 1);
+        } finally {
+            a.recycle();
+        }
+
+        rawData = new ArrayList<>(numDataSets);
+        dataSets = new ArrayList<>(numDataSets);
+        for (int i = 0; i < numDataSets; i++) {
+            rawData.add(new ArrayList<Double>());
+            final BarDataSet dataSet = new BarDataSet(new ArrayList<BarEntry>(), "");
+            dataSet.setColor(ColorTemplate.MATERIAL_COLORS[i]);
+            dataSets.add(dataSet);
+        }
+
+        BarData barData = new BarData(dataSets);
+        barData.setBarWidth(0.45f);
+        setData(barData);
+        groupBars(minBin, 0.08f, 0.01f);
+        final Description desc = new Description();
+        desc.setText(descString);
+        desc.setTextSize(12f);
+        setDescription(desc);
+
+        XAxis xAxis = getXAxis();
+        xAxis.setCenterAxisLabels(true);
+        xAxis.setGranularityEnabled(true);
+        xAxis.setGranularity(1);
+        xAxis.setValueFormatter(new IAxisValueFormatter() {
+            DecimalFormat df = new DecimalFormat("#.##");
+
+            @Override
+            public String getFormattedValue(float value, AxisBase axis) {
+                final float mid = (value - minBin) * BIN_WIDTH + minBin;
+                if (axis.mEntries.length > 1) {
+                    final float width = (axis.mEntries[1] - axis.mEntries[0])*BIN_WIDTH;
+                    return df.format(mid - width/2) + " - " + df.format(mid + width/2);
+                } else {
+                    return df.format(mid);
+                }
+            }
+        });
+
+        invalidate();
+    }
+
+    public void clearData() {
+        for (int i = 0; i < rawData.size(); i++) {
+            rawData.get(i).clear();
+            dataSets.get(i).clear();
+        }
+        invalidate();
+    }
+
+    void addEntry(int dataSetIndex, double value) {
+        if (isRawDataEmpty()) {
+            min = value;
+            max = value;
+        } else {
+            if (value < min) min = value;
+            if (value > max) max = value;
+        }
+
+        rawData.get(dataSetIndex).add(value);
+        minBin = (int) (Math.floor(min/BIN_WIDTH)*BIN_WIDTH);
```

Comment:
Sounds good, I've updated it to be `floor`, and now the labels appear on the boundaries.